adcc/OneParticleOperator.py

```diff
@@ -52,41 +58,201 @@ def __init__(self, spaces, is_symmetric=True):
-        ret = cls(cpp_operator.mospaces, cpp_operator.is_symmetric)
-        for b in cpp_operator.blocks_nonzero:
-            ret.set_block(b, cpp_operator.block(b))
+        occs = sorted(self.mospaces.subspaces_occupied, reverse=True)
+        virts = sorted(self.mospaces.subspaces_virtual, reverse=True)
+        self.orbital_subspaces = occs + virts
+        # check that orbital subspaces are correct
+        assert sum(self.mospaces.n_orbs(ss) for ss in self.orbital_subspaces) \
+            == self.mospaces.n_orbs("f")
+        if self.is_symmetric:
+            # unique combinations of orbital subspaces (respects the symmetry)
+            # and repeated elements (e.g. o1o1)
+            combs = list(
+                combinations_with_replacement(self.orbital_subspaces, r=2)
+            )
+        else:
+            # Cartesian product of subspaces, equivalent to nested for loop
+            combs = list(product(self.orbital_subspaces, repeat=2))
+        self.blocks = ["".join(com) for com in combs]
+        self._tensors = {}
+
+    @property
+    def shape(self):
+        """
+        Returns the shape tuple of the OneParticleOperator
+        """
+        size = self.mospaces.n_orbs("f")
+        return (size, size)
+
+    @property
+    def size(self):
+        """
+        Returns the number of elements of the OneParticleOperator
+        """
+        return np.prod(self.shape)
+
+    @property
+    def blocks_nonzero(self):
+        """
+        Returns a list of the non-zero block labels
+        """
+        return [b for b in self.blocks if b in self._tensors]
+
+    def is_zero_block(self, block):
+        """
+        Checks if block is explicitly marked as zero block.
+        Returns False if the block does not exist.
+        """
+        if block not in self.blocks:
+            return False
+        return block not in self.blocks_nonzero
+
+    def block(self, block):
+        """
+        Returns tensor of the given block.
+        Does not create a block in case it is marked as a zero block.
+        Use __getitem__ for that purpose.
+        """
+        if block not in self.blocks_nonzero:
+            raise KeyError("The block function does not support "
+                           "access to zero-blocks. Available non-zero "
+                           f"blocks are: {self.blocks_nonzero}.")
+        return self._tensors[block]
+
+    def __getitem__(self, block):
+        if block not in self.blocks:
+            raise KeyError(f"Invalid block {block} requested. "
+                           f"Available blocks are: {self.blocks}.")
+        if block not in self._tensors:
+            sym = libadcc.make_symmetry_operator(
+                self.mospaces, block, self.is_symmetric, "1"
+            )
+            self._tensors[block] = Tensor(sym)
+        return self._tensors[block]
+
+    def __getattr__(self, attr):
+        from . import block as b
+        return self.__getitem__(b.__getattr__(attr))
```

Comment:
Nice idea, but if you introduce it, you should also use it: Can you replace old usage of the block strings for this "new" syntax in adcc? There are plenty of places in the equations where one can relplace the `p0_oo` aliases by a single `p0` alias now.